synapse/visibility.py

```diff
@@ -81,8 +81,9 @@ async def filter_events_for_client(
 
     types = ((EventTypes.RoomHistoryVisibility, ""), (EventTypes.Member, user_id))
 
+    # we exclude outliers at this point, and then handle them separately later
     event_id_to_state = await storage.state.get_state_for_events(
-        frozenset(e.event_id for e in events),
+        frozenset(e.event_id for e in events if not e.internal_metadata.outlier),
```

Comment:
it's worth noting that currently, `get_state_for_events` returns an empty list when called on an outlier (which is why this currently works).

However, I'm planning to change it to raise an exception (because I'm pretty sure that doing so is normally a programming error and produces mess like https://github.com/matrix-org/synapse/issues/12074), which is why this PR is needed.